### PR TITLE
[RAD-2440] Unit tests for full GRPC server

### PR DESCRIPTION
**Pull request by mark-idleman**
Removes original server unit tests that were spinning up the old HTTP servers + hitting them with requests, and replaces them with similar tests that operate on the GRPC servers we actually use in production. Next steps will be testing the actual underlying server implementation (instead of using the servers as a black box and testing end-to-end querying + results), ~and of course getting Circle to run these tests [automatically](https://replicahq.atlassian.net/browse/RAD-2548)~.

-  Adds mini_kc test data for use in unit tests. It turns out that mini_kc is _so_ small that very few ODs actually work for transit requests; I've found a specific OD that does return a full-fledge transit route, so I stuck with that for these tests
- Tests added so far are simple, but check quite a lot. I mostly focused on ensuring that all fields in response were included + made sense heuristically (where applicable), with a few tests also checking behavior when ODs are "bad" (outside of the region's bbox). It would also be nice to test "bad" points within the bbox - ones that Graphhopper can snap to the road network, but can't find routes for, but given how small the mini_kc cutout is, I'm not sure they exist here
- GRPC test server was spun up following guidance from [here](https://stackoverflow.com/questions/37552468/how-to-unit-test-grpc-java-server-implementation-functions) and [here](https://grpc.github.io/grpc-java/javadoc/io/grpc/inprocess/InProcessServerBuilder.html), if anyone's curious about the internals
- Automates the running of all tests in CircleCI. Splits up the main dockerfile we use into 2 separate files; one contains all the stages needed to build everything, and gives us access to maven so we can run tests with it; the latter "server" dockerfile runs the final stage that grabs the minimal jars + configs needed to spin up the server, and ends with the `CMD` that runs the server itself. In between these two stages, we run `mvn test` on the first image, as a new step in our CircleCI build

### CODE REVIEW

**Comment by mark-idleman** [comment]
OK, I'm actually trying to get circle to do `mvn test` for us in this PR. One of these attempts is bound to work....(right?)

File: web/src/test/java/com/replica/RouterServerTest.java [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+                .build().start();
+        ManagedChannel channel = InProcessChannelBuilder.forName(uniqueName)
+                .directExecutor()
+                .build();
+
+        routerStub = router.RouterGrpc.newBlockingStub(channel);
+    }
+
+    @BeforeAll
+    public static void setUp() throws Exception {
```
Comment (marcopalmeri) [inline web/src/test/java/com/replica/RouterServerTest.java:157]: @mark-idleman - It might be worth factoring this out into some test utils or something, so it can be reused in other tests?

Comment (mark-idleman) [reply web/src/test/java/com/replica/RouterServerTest.java:157]: Yeah, I think that's definitely a good idea, I was just debating doing it now vs. in the next change because we still don't have very many tests. But, might be nice to just combine now, I'll see how much I can move to a shared test util(s)

**Comment by mark-idleman** [comment]
There are about 1000 ways to get tests _almost_ working in Circle given all the Docker magic we're already doing, but I finally found a way that _actually_ works. Our Circle workflow now has a step where it runs `mvn test` after all of the code is built, which required splitting our current dockerfile into 2 so we can run tests when we still have access to maven in the image we're constructing. There are probably ways to make tests a separate github check (ie a separate Circle job, instead of a step of our current build job), but it's messy. I'll leave it for another day.

**Comment by danielhfrank** [comment]
I think this looks promising! I don't have a lot to say about the application logic of the tests you added, it all seems reasonable to me. A couple of things about the image-publishing changes though:
* can you push up a revision with a change that makes the tests fail, to confirm that the circleci job fails?
* besides the circleci job succeeding, how can we confirm that the router will still run as expected after the docker changes?

File: web/src/test/java/com/replica/RouterServerTest.java (hunk 1 above)
Comment (mark-idleman) [reply web/src/test/java/com/replica/RouterServerTest.java]: Refactored and pulled out all shared logic into a new class that I had the individual tests extend from. This also meant that I moved the existing tests @danielhfrank added recently for the street export to using mini_kc's OSM + GTFS, instead of the original testing files; I'm not removing those files, though, because they could be useful in the future if we want to test weird GTFS edge cases or something

**Comment by mark-idleman** [comment]
@danielhfrank good questions! 

> can you push up a revision with a change that makes the tests fail, to confirm that the circleci job fails?

Definitely! In the last few commits I've pushed I've actually inadvertently done this (eg [here](https://app.circleci.com/pipelines/github/replicahq/graphhopper/622/workflows/a7e2da94-601e-4106-b807-a15a3f11a6dc/jobs/658)), and it's correctly caused the circle job to fail

> besides the circleci job succeeding, how can we confirm that the router will still run as expected after the docker changes?

The easiest way is a bit manual - building a server with the new base image and trying it out. I've [done that](https://cloud.prefect.io/replica/flow-run/b55fd495-7283-4fed-b9f3-bff2aa4c5b1c) with mini nor cal, building a new router that can be spun up with `docker run -it -p 8998:8998 -p 50051:50051 us.gcr.io/model-159019/mininorcal-gh-router-service:bd740266fa8a66db5ecaa61bf8b18f42994998e2`. I can successfully request routes within the GUI, which _should_ be enough to know that both the backend + frontend components were installed correctly during the Docker build (I actually caught a bug this way, that I corrected with my latest commit)
Screen Shot 2021-06-29 at 10 35 31 AM
A slightly better approach (in a future ticket) will be adding a true integration test, that builds a small router using the base docker image constructed at the end of the Circle CI job, and hits it with requests (similar to the tests I added in this PR, but not hitting an internally-instantiated service)

**Review by kaelgreco — APPROVED** [review]
This is so great and it makes me feel 100x better about the repo.

It might make sense to transition to `mini_nor_cal` over `mini_kc` for test data since `mini_kc` is so small and has some idiosyncrasies due to its manual construction. Not necessary now.

It'd also be nice to add basic info to the repo readme to explain testing framework (how to run all /class/single). but it probably makes more sense to do when we include other java build/run info that's currently in the python repo.

it'd be cool to get some semblance of test coverage in place and report +/- on commit too

File: test_gh_config.yaml [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# Configuration file for standard GraphHopper routers.
```
[the remark this answers is not in the document]
Comment (mark-idleman) [reply test_gh_config.yaml:1]: The main difference is the GTFS + OSM are hardcoded in the test config, whereas they're left blank in the default config. With the default config, we can pass the GTFS + OSM paths at runtime using Dropwizard arguments, but I wasn't able to get something similar working with the programatic CLI commands that we run during test setup here, so I did the next best thing and just created a config where they're hardcoded

File: web/src/test/java/com/replica/RouterServerTest.java [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+            observedDistanceMeters += footLeg.getDistanceMeters();
+        }
+        assertEquals(expectedTravelSegmentTypes, observedTravelSegmentTypes);
+        // todo: once PT legs have distances, incorporate those in this check
+        assertEquals(path.getDistanceMeters(), observedDistanceMeters);
+
+        // Check that PT leg contains proper info
+        RouterOuterClass.PtLeg ptLeg = path.getPtLegs(0);
+        assertTrue(ptLeg.getArrivalTime().getSeconds() > ptLeg.getDepartureTime().getSeconds());
+        assertTrue(ptLeg.getStableEdgeIdsCount() > 0); // check that the GTFS link mapper worked
```
[the remark this answers is not in the document]
Comment (mark-idleman) [reply web/src/test/java/com/replica/RouterServerTest.java:181]: 🎉

**Review by JacobHayes — COMMENTED** [review]
Dockerfile changes LGTM. Will we need to update the `model` side to use the new `com.replica.` prefix in the java commands?

File: web/src/test/java/com/replica/RouterServerTest.java [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+import static org.junit.jupiter.api.Assertions.*;
+
+/**
+ * Tests the entire server, not the server implementation itself, so that the plugging-together
+ * of stuff (which is different for PT than for the rest) is under test, too.
+ */
+@ExtendWith(DropwizardExtensionsSupport.class)
+public class RouterServerTest extends ReplicaGraphHopperTest {
+    private static final Timestamp EARLIEST_DEPARTURE_TIME =
+            Timestamp.newBuilder().setSeconds(Instant.parse("2017-07-21T08:25:00Z").toEpochMilli() / 1000).build();
```
[the remark this answers is not in the document]
Comment (mark-idleman) [reply web/src/test/java/com/replica/RouterServerTest.java]: Hmmm good call, I'll add a small note with the official mini_kc validity start date + a hint at the bbox. Actually, we weren't using the most up-to-date date here anyway, so I updated it to match with the current validity start date in mini_kc

**Comment by mark-idleman** [comment]
@kaelgreco 100% agree that `mini_nor_cal` would be better here, and that's what I started with, but the OSM is too big for github to let you push it :( We'd have to set something up so it would be downloaded from GCS at test startup or something, annoying but do-able.

re: your other suggestions on adding info to the readme (eventually in this repo, not model) and getting a more formal report on test coverage/etc, I also completely agree. There's a lot more to add to this basic foundation, but just having tests that run every commit feels very good to start!

@JacobHayes I don't think so! The only thing that was moved to a new package was the server itself, and the only spot that needs updating with the `com.replica` prefix for the server is in the Dockerfile itself, which I've updated in this PR. The other commands haven't seen any changes here, and I guess we got confirmation that they still work when I was able to build mini nor cal completely with a base image from this branch 😁
